util/cache-middleware/index.js

```diff
@@ -0,0 +1,23 @@
+const { validateParams } = require('./api.js');
+const cache = require('./cache.js');
+
+// universal cache handling middleware
+// returns cached data immediately after receiving a request
+// and then then tries to update cache from the DB
+async function cacheMiddleware(req, res) {
+  try {
+    validateParams(req);
+    res.status(200).json({
+      ...cache.get(req),
+    });
+    await cache.update(req);
+  } catch (error) {
+    res.status(400).json({
```

Comment:
before sending the error response as a HTTP resonse, need to check if the the valid response has already been sent within the try. (because the `throw` can occur during `await cache.update(req);`.

if this occurs, then the error should be handled by logging, instead of a HTTP response.

Reply:
I anticipated this, so update never throws errors. It is run inside a try block.

Reply:
I did it deliberately because I wanted no action if there would be any error during the attempt of updating the cache . Could not update the cache - I am OK with this. I will be trying to update it again within a next API request

Reply:
anyway, I added a check for sent headers

Reply:
ah OK, if `cache.update` never throws then no need changes here.